api/lib/opentelemetry/trace/propagation/trace_context/text_map_propagator.rb

```diff
@@ -45,8 +45,12 @@ def inject(carrier, context: Context.current, setter: Context::Propagation.text_
           # @return [Context] context updated with extracted baggage, or the original context
           #   if extraction fails
           def extract(carrier, context: Context.current, getter: Context::Propagation.text_map_getter)
-            tp = TraceParent.from_string(getter.get(carrier, TRACEPARENT_KEY))
-            tracestate = Tracestate.from_string(getter.get(carrier, TRACESTATE_KEY))
+            trace_parent_value = getter.get(carrier, TRACEPARENT_KEY)
+            trace_state_value = getter.get(carrier, TRACESTATE_KEY)
+            return context if trace_parent_value.nil? || trace_state_value.nil?
+
+            tp = TraceParent.from_string(trace_parent_value)
+            tracestate = Tracestate.from_string(trace_state_value)
```

Comment:
I believe I understand the intention of this PR now.

This change is intends to stop relying on exceptions for control flow in the (potentially common) case where the TRACEPARENT key is not present in the carrier, so instead we return the context early if we failed to extract any value from that key.

So this is not resolving any bug but instead meant to improve performance, is that correct @hosamaly?

If we do take this approach, we should *not* return early if the trace_state_value is nil, as the `Tracestate.from_string` can handle that case.  

```suggestion
            tracestate = Tracestate.from_string(getter.get(carrier, TRACESTATE_KEY))
            return context unless trace_parent_value = getter.get(carrier, TRACEPARENT_KEY)
            tp = TraceParent.from_string(trace_parent_value)
```

Reply:
Unfortunately, we can't parse `tracestate` before parsing `traceparent`:
> If the vendor failed to parse `traceparent`, it _MUST NOT_ attempt to parse `tracestate`.

Reply:
Indeed, you're absolutely right. I've committed your suggestion and reworded the commit to indicate that it's a performance improvement rather than a fix.